### ISSUE TITLE
Support for maps out of the box?

### ISSUE THREAD

**Issue by krypton97** [jackc/pgx#245]
Hey, I've recently started to play with graphql api and decided to use this library (which is truely a pleasure to use) to connect with psql. I'm curious if there's a way to recieve the data as custom map type. So far I'm fetching the data from the db and convert it to what I'm looking for. This is what I got working so far:

	func(p *graphql.ResolveParams) (interface{}, error) {
				rows, _ := db.Query("SELECT * FROM courses;")
				var resolver []map[string]interface{}
				for rows.Next() {
					var a, b, c, d, e string
					m := make(map[string]interface{})
					_ = rows.Scan(&a, &b, &c, &d, &e)

				m["cursor"] = a
				m["node"] = map[string]interface{}{
					"id":        b,
					"title":     c,
					"url":       d,
					"createdAt": e,
				}
				resolver = append(resolver, m)
			}

			return resolver, nil
		}

What I got so far works good, didn't implement yet all the features, but also it would be nice to know if that's the way rigth way to work with maps, ty!

**Comment by jackc** [comment]
As long as you know in advance what columns you are expecting back that is a reasonable approach. If you don't know what columns to expect ahead of time, then [*Rows.FieldDescriptions](https://godoc.org/github.com/jackc/pgx#Rows.FieldDescriptions) and [*Rows.Values](https://godoc.org/github.com/jackc/pgx#Rows.Values) can be used to get the field names and values.

**Comment by krypton97** [comment]
I've changed my mind and I decided to insert directly as JSON string like ` {
            "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
            "node": {
              "createdAt": "1974-08-07T11:53:39.055Z",
              "id": "58bc29705d58db311b3d1d85",
              "title": "test link",
              "url": "test.com"
            }
          },`

as a []byte type and when a I fetch it back, to unmarshal it to `map[string]interface{}`, but I also need to do some pattern matching on title or cursor so in this case a I should insert the as varchar separately. Other than that ty for help, I think I'm gonna do a small benchmark to see what fits the best ;)

**Comment by krypton97** [comment]
Gonna post here a small benchmark I've done:
Setup 2 tables `json` and `courses`:

	create table json(
		node varchar(255)
	)
	
	insert into json values('{"cursor":"YXJyYXljb25uZWN0aW9uOjC","node":{"id":"507f191e810c19729de860ec","title":"ReactJS Course","url":"facebook.github.com/graphql/","createdAt":"Mon Jan 2 15:04:05"}}') 
	
	//insert how many times u need (I did 15 insertions)
	//second table 
	
	create table courses(
		cursor varchar(255),
		id varchar(255) not null,
		title varchar(255),
		url varchar(255),
		createdAt varchar(255)
	)
	
	INSERT INTO courses VALUES('YXJyYXljb25uZWN0aW9uOjC','507f191e810c19729de860ec', 'ReactJS Course', 'facebook.github.com/graphql/','Mon Jan 2 15:04:05');
	//again I did 15 insertions like the one above


Now the benchmark where I perform a small regexp:

	
	package graphql_test
	
	import (
		"testing"
	
		"github.com/a8m/djson"
		"github.com/jackc/pgx"
	)
	
	func BenchmarkJSON(b *testing.B) {
		cfgx, _ := pgx.ParseURI("postgres://postgres:postgres@localhost:5432/postgres?sslmode=disable")
		db, _ := pgx.Connect(cfgx)
		defer db.Close()
		for i := 0; i < b.N; i++ {
			rows, _ := db.Query("SELECT * FROM json where node like '%JS Co%'")
			var resolver []map[string]interface{}
			for rows.Next() {
				var obj []byte
				_ = rows.Scan(&obj)
				m, _ := djson.Decode(obj)
				resolver = append(resolver, m.(map[string]interface{}))
			}
		}
	}
	
	func BenchmarkNormal(b *testing.B) {
		cfgx, _ := pgx.ParseURI("postgres://postgres:postgres@localhost:5432/postgres?sslmode=disable")
		db, _ := pgx.Connect(cfgx)
		defer db.Close()
		for i := 0; i < b.N; i++ {
			rows, _ := db.Query("SELECT * FROM courses where title like '%JS Co%'")
			var resolver []map[string]interface{}
			for rows.Next() {
				var a, b, c, d, e string
				m := make(map[string]interface{})
				_ = rows.Scan(&a, &b, &c, &d, &e)
				m["cursor"] = a
				m["node"] = map[string]interface{}{
					"id":        b,
					"title":     c,
					"url":       d,
					"createdAt": e,
				}
				resolver = append(resolver, m)
			}
		}
	}


TL;DR:

	BenchmarkJSON-8            10000            119428 ns/op
	BenchmarkNormal-8          10000            123913 ns/op

 The `json` takes only 8192B, the `course` table takes 16kB which is pretty amazing to see how big is the size difference for almost the same performance ;). This a simple-ish case, but it's very good to consider it when working with maps

**Comment by jackc** [comment]
Interesting. A few ideas you might want to look into: PostgreSQL has `json` and `jsonb` types. If you want to store JSON in the database I suggest using them instead of `varchar`. With `jsonb` you even get indexing of JSON documents.

Beyond that, if your data has a structure that is known ahead of time you are usually better off using relational structure rather than storing it in a text or json blob. If you need to output JSON PostgreSQL has support for [json generation in SQL](
https://hashrocket.com/blog/posts/faster-json-generation-with-postgresql).

**Comment by krypton97** [comment]
You're absolutely right, figured it out that psql is better to use as a relation db, json kinda slows it bit. Thank you for the support :)
